metadata-io/src/main/java/com/linkedin/metadata/trace/KafkaTraceReader.java

```diff
@@ -0,0 +1,460 @@
+                      tp.topic(),
+                      tp.partition(),
+                      consumerGroupId);
+                  return null;
+                }
+
+                log.debug(
+                    "Found offset metadata {} for Topic: {}, Partition: {}, Group: {}",
+                    offsetAndMetadata,
+                    tp.topic(),
+                    tp.partition(),
+                    consumerGroupId);
+
+                return offsetAndMetadata;
+              } catch (SchemaException e) {
+                log.error("Schema error when fetching consumer group offsets", e);
+                return null;
+              } catch (Exception e) {
+                log.error("Error fetching consumer group offsets", e);
+                return null;
+              }
+            }));
+  }
+
+  private Map<String, Pair<ConsumerRecord<String, GenericRecord>, SystemMetadata>> findMessages(
+      Urn urn, Collection<String> aspectNames, String traceId, Long traceTimestampMillis)
+      throws ExecutionException, InterruptedException {
+
+    TopicPartition topicPartition = getTopicPartition(urn);
+
+    try (Consumer<String, GenericRecord> consumer = consumerSupplier.get()) {
+      // Assign the partition we want to read from
+      consumer.assign(Collections.singleton(topicPartition));
+
+      // Get offset for timestamp
+      OffsetAndTimestamp offsetAndTimestamp =
+          getOffsetByTime(consumer, topicPartition, traceTimestampMillis);
+
+      if (offsetAndTimestamp == null) {
+        log.debug(
+            "No offset found for timestamp {} in partition {}",
+            traceTimestampMillis,
+            topicPartition);
+        return Collections.emptyMap();
+      }
+
+      // Seek to the offset for the timestamp
+      consumer.seek(topicPartition, offsetAndTimestamp.offset());
+      log.debug(
+          "Seeking to timestamp-based offset {} for partition {}",
+          offsetAndTimestamp.offset(),
+          topicPartition);
+
+      // Poll with a maximum number of attempts
+      int attempts = 0;
+      long lastProcessedOffset = -1;
+      Map<String, Pair<ConsumerRecord<String, GenericRecord>, SystemMetadata>> results =
+          new HashMap<>();
+
+      while (attempts < pollMaxAttempts && results.size() < aspectNames.size()) {
+        var records = consumer.poll(java.time.Duration.ofMillis(pollDurationMs));
+        attempts++;
+
+        if (records.isEmpty()) {
+          break;
+        }
+
+        // Check if we're making progress
+        long currentOffset = consumer.position(topicPartition);
+        if (currentOffset == lastProcessedOffset) {
+          break;
+        }
+        lastProcessedOffset = currentOffset;
+
+        // Process records for each aspect name we haven't found yet
+        for (String aspectName : aspectNames) {
+          if (!results.containsKey(aspectName)) {
+            var matchingRecord =
```

Comment:
We generally avoid var in the rest of the codebase so a bit weird to see it here, but I think it's fine since we don't need to run metadata-io in Java 11/8 compat